h2o-algos/src/main/java/hex/glm/ComputationState.java

```diff
@@ -925,6 +1103,247 @@ protected void setHGLMComputationState(double [] beta, double[] ubeta, double[]
+      _gram = gramM;
+      beta = b;
+      _grad = grad;
+      objective = obj;
+      _sumOfRowWeights = sumOfRowWeights;
+      _xy = xy;
+    }
+
+    public Gram.Cholesky cholesky(Gram.Cholesky chol, double[][] xx) {
+      if( chol == null ) {
+        for( int i = 0; i < xx.length; ++i )
+          xx[i] = xx[i].clone();
+        chol = new Gram.Cholesky(xx, new double[0]);
+      }
+      final Gram.Cholesky fchol = chol;
+      final int sparseN = 0;
+      final int denseN = xx.length - sparseN;
+      // compute the cholesky of the diagonal and diagonal*dense parts
+      ForkJoinTask [] fjts = new ForkJoinTask[denseN];
+      // compute the outer product of diagonal*dense
+      //Log.info("SPARSEN = " + sparseN + "    DENSEN = " + denseN);
+      final int[][] nz = new int[denseN][];
+      for( int i = 0; i < denseN; ++i ) {
+        final int fi = i;
+        fjts[i] = new RecursiveAction() {
+          @Override protected void compute() {
+            int[] tmp = new int[sparseN];
+            double[] rowi = fchol._xx[fi];
+            int n = 0;
+            for( int k = 0; k < sparseN; ++k )
+              if (rowi[k] != .0) tmp[n++] = k;
+            nz[fi] = Arrays.copyOf(tmp, n);
+          }
+        };
+      }
+      ForkJoinTask.invokeAll(fjts);
+      for( int i = 0; i < denseN; ++i ) {
+        final int fi = i;
+        fjts[i] = new RecursiveAction() {
+          @Override protected void compute() {
+            double[] rowi = fchol._xx[fi];
+            int[]    nzi  = nz[fi];
+            for( int j = 0; j <= fi; ++j ) {
+              double[] rowj = fchol._xx[j];
+              int[]    nzj  = nz[j];
+              double s = 0;
+              for (int t=0,z=0; t < nzi.length && z < nzj.length; ) {
+                int k1 = nzi[t];
+                int k2 = nzj[z];
+                if (k1 < k2) { t++; continue; }
+                else if (k1 > k2) { z++; continue; }
+                else {
+                  s += rowi[k1] * rowj[k1];
+                  t++; z++;
+                }
+              }
+              rowi[j + sparseN] = xx[fi][j + sparseN] - s;
+            }
+          }
+        };
+      }
+      ForkJoinTask.invokeAll(fjts);
+      // compute the cholesky of dense*dense-outer_product(diagonal*dense)
+      double[][] arr = new double[denseN][];
+      for( int i = 0; i < arr.length; ++i )
+        arr[i] = Arrays.copyOfRange(fchol._xx[i], sparseN, sparseN + denseN);
+      final int p = H2ORuntime.availableProcessors();
+      Gram.InPlaceCholesky d = Gram.InPlaceCholesky.decompose_2(arr, 10, p);
+      fchol.setSPD(d.isSPD());
+      arr = d.getL();
+      for( int i = 0; i < arr.length; ++i ) {
+        // See PUBDEV-5585: we use a manual array copy instead of System.arraycopy because of behavior on Java 10
+        // Used to be: System.arraycopy(arr[i], 0, fchol._xx[i], sparseN, i + 1);
+        for (int j = 0; j < i + 1; j++)
+          fchol._xx[i][sparseN + j] = arr[i][j];
+      }
```

Comment:
Do we think the problem with `System.arraycopy` is limited to this piece of code? If not, maybe we should consider moving it to e.g. `ArrayUtils` and make sure we use it elsewhere as well.  (Something like: check if we are running on java 10, do specific logic, otherwise use `System.arraycopy`).

Reply:
This change was made by Michal K.  I don't know why it is not working.

Reply:
I read the jira ticket and the problem was with Java 10 + JIT compilation and it was hard to reproduce. I think we can disregard my comment as it might be already fixed in Java 10.